### PR TITLE
Add C++ API for `set_default_enabled` to globally enable/disable recording streams.

### PR DESCRIPTION
**Pull request by jleibs**
### What
Resolves:
 - https://github.com/rerun-io/rerun/issues/3754

Adds a very minimal singleton config system and an API to toggle the state of default_recordings. Specifically this makes it possible for a user to globally disable default recordings, and then enable them via the RERUN environment variable.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/4014) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4014)
- [Docs preview](https://rerun.io/preview/3144bc8776ec355d65c5ccc94d8bac385662f926/docs) 
- [Examples preview](https://rerun.io/preview/3144bc8776ec355d65c5ccc94d8bac385662f926/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: crates/rerun_c/src/lib.rs [hunk 1]
```diff
@@ -52,6 +52,8 @@ pub struct CStoreInfo {
     pub application_id: *const c_char,
 
     pub store_kind: CStoreKind,
+
+    pub default_enabled: bool,
```
Comment (teh-cmc) [inline crates/rerun_c/src/lib.rs:56]: This doesn't belong here, it breaks the fact that `CStoreInfo` is a mirror of `StoreInfo`... Do you really need to store this here?

File: rerun_cpp/src/rerun/config.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            );
+            if (envVarValueStr == "1" || envVarValueStr == "true" || envVarValueStr == "yes") {
+                default_enabled.store(true, std::memory_order_seq_cst);
+            } else if (envVarValueStr == "0" || envVarValueStr == "false" || envVarValueStr == "no") {
+                default_enabled.store(false, std::memory_order_seq_cst);
+            }
+        } else {
+            default_enabled = true;
+        }
+    }
```
Comment (teh-cmc) [inline rerun_cpp/src/rerun/config.cpp:35]: I'm confused. If the whole point of the `RERUN` var is to override the value of `default_enabled`, then why are setting `default_enabled` to the value of the `RERUN` var...? What even is the point of `default_enabled` then?

I'd expect `default_enabled` to simply always be `true` unless the user changes it from code, like in other SDKs (I think?).

File: rerun_cpp/src/rerun/config.hpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    inline void set_default_enabled(bool default_enabled) {
+        RerunGlobalConfig::instance().default_enabled.store(
+            default_enabled,
+            std::memory_order_seq_cst
+        );
+    }
+
+    /// Check if Rerun is enabled.
+    inline bool is_default_enabled() {
+        return RerunGlobalConfig::instance().default_enabled.load(std::memory_order_relaxed);
```
Comment (teh-cmc) [inline rerun_cpp/src/rerun/config.hpp:43]: Honestly, this being C++, I'd just make it all `seq_cst`... one less thing to worry about and it's not like we're fighting for contention here.

Comment (Wumpf) [reply rerun_cpp/src/rerun/config.hpp:43]: I don't disagree with the sentiment, but seq_cst can be quite expensive on arm though and we want to do this thing in a lot of places.
If another thread changes the enabled state there's no way anyways to determine when that happens in relation to instruction on the other state. Afaik the only thing that this protects us from is that instructions that in some cases the code "look like they come later" may see a different enabled state than those that "look like they come earlier" (as long as it doesn't break logic on a single thread).

Comment (teh-cmc) [reply rerun_cpp/src/rerun/config.hpp:43]: > we want to do this thing in a lot of places.

That doesn't sound right? `is_default_enabled` should only be relevant when creating a recording stream, which is quite rare.

File: rerun_cpp/src/rerun/config.cpp (hunk 2 above)
Comment (jleibs) [reply rerun_cpp/src/rerun/config.cpp:35]: I guess I can drop this -- there was some historical context where this whole config option was originally `set_enabled` as described in [the](https://github.com/rerun-io/rerun/issues/3754) and implemented before we remembered that rust was going to independently parse the RERUN var.

File: rerun_cpp/src/rerun/config.hpp (hunk 3 above)
Comment (jleibs) [reply rerun_cpp/src/rerun/config.hpp:43]: Yeah, again, the usage of atomic here stems from when we thought this check was going to happen inside every log call as opposed to just during stream construction.

File: crates/rerun_c/src/lib.rs (hunk 1 above)
Comment (jleibs) [reply crates/rerun_c/src/lib.rs]: Got it -- we weren't exactly using CStoreInfo as a mirror of StoreInfo in any meaningful way given we just deconstruct it to pass the arguments into the builder. I had just interpreted it as something more like "RecordingStreamConstructionArgs". But I'll just move it to a function argument instead.

**Review by teh-cmc — APPROVED** [review]

**Pull request merged**